src/create-context.js

```diff
@@ -21,18 +21,21 @@ export function createContext(defaultValue) {
 
 				this.shouldComponentUpdate = _props => {
 					if (this.props.value !== _props.value) {
-						subs.some(c => {
-							c.context = _props.value;
-							enqueueRender(c);
+						subs.some(payload => {
+							payload[0].context = _props.value;
+							if (!payload[1] || payload[1](_props.value, this.props.value)) {
```

Comment:
payload[1] can be a closure over props from a pervious render. if the author isn't careful the shouldUpdate function may throw since the potentially soon to be stale props won't be compatible with the newly updating value.

I'm not super familiar with the preact codebase so apologies if this is just a distraction and error recovery is handled elsewhere that would protect against this case

Reply:
That's an issue that can only occur in 2phase renderers if I'm not mistaking

Reply:
Hey, I've added tests for this

https://github.com/preactjs/preact/pull/2452/files#diff-507a7cc69be8471ba7e6c38f43a7cd4eR437

Reply:
the update looks good and solves the more severe case of stale props (basically capturing it on first use and never updating it looks like) but I was actually more thinking of this edge case

Imagine this component
```
function MyComponent({ someKey }) {
  // update when name changes from dynamically referenced context property
  let ctx = useContext(MyContext, (next, prev) => { next[someKey].name !== prev[someKey] .name})
  return ... // use ctx[someKey].name presumably
}
```

```
let things = {
  aThing: { name: 'foo' }
};

//render
<MyContext.Provider value={things}>
  <MyComponent someKey={Object.keys(things)[0]} /> //get the first thing key
</MyContext.Provider>

then change things
things = {
  aDifferentThing: { name: 'bar' }
};

//render again
// when Provider calls subs the captured update will do things[aThing].name and will throw
<MyContext.Provider value={things}>
  <MyComponent someKey={Object.keys(things)[0]} />
</MyContext.Provider>
```